### ISSUE TITLE
Snapshot path for snapshots created with stdin

### ISSUE THREAD

**Issue by jrouleau** [restic/restic#1809]
## Output of `restic version`
restic 0.9.0 (v0.9.0-1-g20bfed59) compiled with go1.10.2 on linux/amd64

## How did you run restic exactly?
```
$ restic -r /tmp/repo -p /tmp/repo_password init
[...]
$ echo foo | restic -r /tmp/repo -p /tmp/repo_password backup --stdin --stdin-filename foo
[...]
$ restic -r /tmp/repo -p /tmp/repo_password ls latest
repository fd5bd830 opened successfully, password is correct
snapshot 8d9d9ac2 of [/home/jonathan/foo] at 2018-05-24 06:00:55.208260662 +0000 UTC):
/foo
```

## What backend/server/service did you use to store the repository?
local

## Expected behavior
The snapshot path (`snapshot <id> of [<path>]`) for snapshots created with stdin should be sensible and not misleading.

## Actual behavior
The snapshot path for stdin snapshots is `<working dir>/<stdin filename>` which likely points to a filepath that never existed which can cause confusion and does not accurately represent the source of the data. 

## Steps to reproduce the behavior
_see above_

## Do you have any idea what may have caused this?
https://github.com/restic/restic/blob/v0.9.0/internal/restic/snapshot.go#L35

> Abs returns an absolute representation of path. **If the path is not absolute it will be joined with the current working directory to turn it into an absolute path**. The absolute path name for a given file is not guaranteed to be unique. Abs calls Clean on the result.
[golang.org](https://golang.org/pkg/path/filepath/#Abs)

## Do you have an idea how to solve the issue?
- Use `/dev/stdin` as the source path. This should not affect how the file is stored within the snapshot (`/<stdin filename>`)

**Comment by fd0** [comment]
Oh, interesting, that's a bug. Thanks for the report!

**Comment by Angristan** [comment]
Same with `restic 0.9.1 compiled with go1.10.2 on linux/amd64`, from the Debian Sid package.

**Comment by fd0** [comment]
Yep, it's not fixed (yet).

**Comment by whereisaaron** [comment]
Thanks @jrouleau I was coming to report the same issue.

I was actually hoping the `--stdin-filename` would accept the whole file path to use, e.g. `--stdin-filename=/foo/bar/filename.sql.gz`. When I tried that out with 0.9.1 I got a weird `lstat` file not found error. So might be a logic issue there too.

**Comment by fd0** [comment]
Both issues are related, and now with the new archiver code (since 0.9.0) the issues can be resolved. There just needs to be someone actually doing it (or we'll just wait until I come around to do it) :)

**Comment by mrmachine** [comment]
So is it currently impossible to use `--stdin` with S3?

**Comment by angristan** [comment]
@mrmachine I'm currently using restic with an S3 backend (wasabi) with the `--stdin` option, if that's what you're asking.

**Comment by fd0** [comment]
> So is it currently impossible to use --stdin with S3?

Not at all, it works very well. The path listed in the overview displayed by `restic snapshots` just has the current directory prefixed to the string `stdin`, that's all. It's just a minor annoyance, no functional limitations involved.

**Comment by mrmachine** [comment]
So I just can't use `--stdin` *and* `--stdin-filename` with S3?

**Comment by fd0** [comment]
> So I just can't use --stdin and --stdin-filename with S3?

You can do that, but if you run `restic` while the current directory is `/home/user`, using `--stdin --stdin-filename foo.sql`, then the path displayed in the `snapshots` output will be `/home/user/foo.sql`. That's what this bug is about. The path is not even used for anything.

If you still find that annoying, just run `cd /` before running restic, so the path prefixed to the filename given in `--stdin-filename` will be `/`.

**Comment by mrmachine** [comment]
Sorry, I was getting an lstat file not found error (same as @whereisaaron above) that I thought was caused by this issue.

**Comment by whereisaaron** [comment]
The `lstat` error is mentioned in #2063 and a possible fix in #2206.
